opal/mca/common/ofi/common_ofi.c

```diff
@@ -446,63 +446,116 @@ static int check_provider_attr(struct fi_info *provider_info, struct fi_info *pr
- *     @param (IN) pci              struct fi_pci_attr pci device attributes,
- *                                  used to find hwloc object for device.
- *
- *     @param (IN) topology         hwloc_topology_t topology to get the cpusets
- *                                  from
- *
- *     @param (OUT)                 returns true if cpusets match and false if
- *                                  cpusets do not match or an error prevents comparison
- *
- *     Uses a pci device to find an ancestor that contains a cpuset, and
- *     determines if it intersects with the cpuset that the process is bound to.
- *     if the process is not bound, or if a cpuset is unavailable for whatever
- *     reason, returns false. Otherwise, returns the result of
- *     hwloc_cpuset_intersects()
- */
-static bool compare_cpusets(hwloc_topology_t topology, struct fi_pci_attr pci)
+
+#define MAX_NUM_NUMA 8
+static pmix_device_distance_t *get_nearest_nics(int *num_distances)
 {
-    bool result = false;
-    int ret;
+	size_t ndist;
+	pmix_value_t *val;
+	int ret, i, idx = 0;
+	pmix_cpuset_t cpuset;
+	pmix_topology_t *topo;
+	pmix_data_array_t *dptr;
+	uint16_t near = USHRT_MAX;
     hwloc_bitmap_t proc_cpuset;
-    hwloc_obj_t obj = NULL;
-
-    /* Cannot find topology info if no topology is found */
-    if (NULL == topology) {
-        return false;
-    }
-
-    /* Allocate memory for proc_cpuset */
-    proc_cpuset = hwloc_bitmap_alloc();
-    if (NULL == proc_cpuset) {
-        return false;
-    }
-
-    /* Fill cpuset with the collection of cpu cores that the process runs on */
-    ret = hwloc_get_cpubind(topology, proc_cpuset, HWLOC_CPUBIND_PROCESS);
-    if (0 > ret) {
-        goto error;
-    }
-
-    /* Get the pci device from bdf */
-    obj = hwloc_get_pcidev_by_busid(topology, pci.domain_id, pci.bus_id, pci.device_id,
-                                    pci.function_id);
-    if (NULL == obj) {
-        goto error;
-    }
+	pmix_info_t *info, directive;
+	pmix_device_distance_t *distances, *nearest = NULL;
+	pmix_device_type_t type = PMIX_DEVTYPE_OPENFABRICS |
+	  PMIX_DEVTYPE_NETWORK |
+	  PMIX_DEVTYPE_COPROC |
+	  PMIX_DEVTYPE_GPU |
+	  PMIX_DEVTYPE_BLOCK |
+	  PMIX_DEVTYPE_DMA;
+
+	PMIX_INFO_LOAD(&directive, PMIX_OPTIONAL, NULL, PMIX_BOOL);
+	ret = PMIx_Get(&opal_process_info.myprocid,
+			 PMIX_DEVICE_DISTANCES, &directive, 1, &val);
+	PMIX_INFO_DESTRUCT(&directive);
+	if (ret != PMIX_SUCCESS) {
+		goto out;
+	}
+
+    if (PMIX_DATA_ARRAY != val->type) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    dptr = val->data.darray;
+    if (NULL == dptr) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    if (PMIX_DEVICE_DIST != dptr->type) {
+        PMIX_VALUE_RELEASE(val);
+		goto release;
+    }
+    distances = (pmix_device_distance_t*)dptr->array;
+
+	for (i = 0; i < dptr->size; i++)
+		fprintf(stderr, "%d: %d:%s:%d:%d\n", getpid(), i, distances[i].uuid,
```

Comment:
Shame on me - I missed the pmix_device_distances_t struct when providing "pretty-print" functions. I'll address that soon.

Reply:
Can you expand why this wouldn't work for other HW?